### ISSUE TITLE
sql: prior extant secondary index with PK in STORING causes subsequent v2.1 ALTER to fail

### ISSUE THREAD

**Issue by dendon** [cockroachdb/cockroach#32186]
**Describe the problem**
**To Reproduce**

After upgrading the database to v2.1.0 (from v2.0.x), it seems that all existing tables with indexes which were created to explicitly include the primary can now not be changed anymore (at all), as every change returns with the message that there is an index including the primary explicitly while it is already included implicitly.

command example:
```ALTER TABLE video ADD COLUMN name_escaped STRING NOT NULL DEFAULT '' CREATE IF NOT EXISTS FAMILY family_name_escaped;```

example message:
pq: index "name_storing_id" already contains column "id"
DETAIL: column "id" is part of the primary index and therefore implicit in all indexes

It seems that it would be possible to drop all affected indexes on every table, yet, they can not be replaced before with correct indexes, which would render the system unusable for the time being.

**Comment by vivekmenezes** [comment]
@dendon can you provide us with the schema so we can reproduce this problem? Thanks

**Comment by dendon** [comment]
Hey @vivekmenezes.

Sure, create the table in v2.0.x with the specified index and then upgrade the database to v2.1.0.
Once done, try adding a column or an index and you will see the message I mentioned.

```
CREATE TABLE video
(
  id SERIAL PRIMARY KEY
);

ALTER TABLE video ADD COLUMN name STRING NOT NULL DEFAULT '' CREATE IF NOT EXISTS FAMILY family_name;
ALTER TABLE video ADD COLUMN name_escaped STRING NOT NULL DEFAULT '' CREATE IF NOT EXISTS FAMILY family_name_escaped;

CREATE INDEX name_storing_id ON video (name) STORING (id);
```

**Comment by knz** [comment]
@jordanlewis I recall you were looking in this area not too long ago to prevent columns from being listed twice. Does this ring a bell?

**Comment by knz** [comment]
Oh actually I see what's happening already!

@dendon please accept our apologies for not seeing this earlier.

The situation is quite simple really: every secondary index already includes implicitly all the columns from the primary key. This is done because the PK columns are needed to look up a row from the secondary. (In contrast to PostgreSQL, CockroachDB does not identifies rows using a OID; instead, the primary key is used to identify rows.)

In addition we changed CockroachDB recently to refuse specifying a column in STORING that is already otherwise stored in the index. This is why you seemingly cannot add a PK column via STORING -- it's already "in there" implicitly. 

You can reveal this by doing the following:

```sql
> CREATE INDEX name_storing_id ON video (name); -- don't use STORING
> SHOW INDEXES FROM video;

  table_name |   index_name    | non_unique | seq_in_index | column_name | direction | storing | implicit
+------------+-----------------+------------+--------------+-------------+-----------+---------+----------+
  video      | primary         |   false    |            1 | id          | ASC       |  false  |  false
  video      | name_storing_id |    true    |            1 | name        | ASC       |  false  |  false
  video      | name_storing_id |    true    |            2 | id          | ASC       |  false  |   true

```


See how `id` is already there "implicitly".

Does this help?

As a side question, can you inform us how you came to this idea? Is this SQL DDL created by some automated tool? If you think this restriction on STORING is limiting compatibility with your client app, we would love to look at this further.

Thank you!

**Comment by knz** [comment]
@jordanlewis (Side note:) IMHO we should *accept* duplicate columns in STORING (or the list of columns in the index spec) and simply ignore them, instead of throwing an error. Also I would suggest that if a PK column was listed in the spec or STORING (even if it would have been added to the index otherwise), it should be marked as non-implicit somehow.

**Comment by dendon** [comment]
@knz I understood the message, that was actually not a question about what is going on, rather if that's actually intended behaviour as it is blocking every kind of future change to solve it.

As in the example there is a name_storing_id index and as long as this exists, running:
```
CREATE INDEX name ON video (name);
```
fails and would only succeed when running before:
```
DROP INDEX video@name_storing_id;
```
But this would lead to the problem that every query runs non-indexed for the time being.


It's actually not a problem, if the primary is always included i could happy discard it from every index.
Though, the problem is that I now would have to first drop every affected index and then recreate it without the storing clause, which is a problem, because I noticed it AFTER upgrading the production system.
So I now have a db sitting there with a few billion entries which I would have to take offline for a week to clean that up.

I am actually building my own CMS, so compatibility is of no concern. I am used to being as explicit as possible and I didn't expected that to be a problem actually.

**Comment by dendon** [comment]
I am sorry if that wasn't clear before, but correct @knz , any change (CREATE INDEX, ALTER TABLE) to a table containing such an index fails with that message.
With the only exception being dropping the 'incorrect' indexes.

**Comment by knz** [comment]
Yeah ok thanks now I see.

@jordanlewis it looks like the change in #31019 must be amended and the followup also backported to 2.0 and 2.1.

**Comment by dendon** [comment]
Hey,

don't wanna rush you or anything, but would it be possible to get an eta?

**Comment by knz** [comment]
@dendon sorry for the late notice. We are prioritizing this. We are aiming for a fix in v2.1.2.

**Comment by knz** [comment]
The ETA for that would be some time in early December.

**Comment by jordanlewis** [comment]
This was closed in November in #32686.
